Setup pre-commit hooks

samderanova opened the pull request:
This PR uses [`pre-commit`](https://www.npmjs.com/package/pre-commit) to set up pre-commit hooks for the following:

- `control-station/`
    - Prettier
    - ESLint
    - `tsc`
 - `pod-operation/`
    - `cargo fmt`
    - `cargo check`
    - `cargo clippy`

samderanova committed:
Add pre-commit and a few basic hooks

samderanova committed:
Add additional pre-commit hooks

samderanova committed:
Add tsc pre-commit hook

taesungh commented on package.json:
Chore: Could we reformat this to tabs

taesungh commented on package.json:
```diff
@@ -0,0 +1,23 @@
+  "version": "1.0.0",
+  "description": "The control system consists of the pod operation backend and control station client. See the respective README files for setup instructions.",
+  "main": "index.js",
+  "dependencies": {
+    "husky": "^9.0.11"
+  },
```
Chore: husky is unused?

samderanova replied:
Will remove

taesungh commented on package.json:
```diff
@@ -0,0 +1,23 @@
+    "pre-commit": "^1.2.2"
+  },
+  "scripts": {
+    "tsc": "cd control-station && tsc --noEmit && exit 0",
+    "eslint": "cd control-station && npx eslint . && exit 0",
+    "prettier": "cd control-station && prettier --write . && exit 0",
+    "cargoclippy": "cd pod-operation && cargo clippy && exit 0",
+    "cargocheck": "cd pod-operation && cargo check && exit 0",
+    "cargofmt": "cd pod-operation && cargo fmt && exit 0"
```
Suggestion: could we reorganize these into logical categories per stack and delegate to the associated app? For example, have a `format` and `lint` script here which delegates to subsequent matching scripts in `control-station/package.json`, similar to a monorepo pipeline.

taesungh commented on package.json:
```diff
@@ -0,0 +1,23 @@
+    "cargocheck": "cd pod-operation && cargo check && exit 0",
+    "cargofmt": "cd pod-operation && cargo fmt && exit 0"
+  },
+  "pre-commit": ["prettier", "cargofmt"],
```
Question: do we need to add the other scripts to this array?

samderanova replied:
Yes, I will be adding them.

samderanova committed:
Miscellaneous fixes:

- Update .prettierrc
- Add pre-commit scripts to run
- Remove unused dependency husky

taesungh reviewed (changes requested):
Nice start, mainly one suggestion about structuring the npm scripts to use corresponding scripts for each app.

samderanova committed:
Reorganize pre-commit hooks into subrepos

samderanova committed:
Use tabs with width of 4 to format package.json

samderanova committed:
Remove unecessary package.json elements, add author

samderanova committed:
Update README.md with pre-commit hooks section

taesungh commented on package.json:
```diff
@@ -0,0 +1,24 @@
+		"format:po",
+		"lint:cs",
+		"lint:po",
+		"test:po"
+	],
```
Chore: think we also need to include `"test:cs"` here?

samderanova replied:
Sure, but do we have any tests for the frontend?

samderanova replied:
Oh wait never mind

taesungh commented on .gitignore:
```diff
@@ -0,0 +1 @@
+node_modules/
```
Nitpick: could we get a new line at EOF here

samderanova committed:
Add test hook for control-station

samderanova committed:
Add test hook for control-station

SM-dot commented on control-station/package.json:
```diff
@@ -7,6 +7,8 @@
 		"dev": "vite",
 		"build": "tsc && vite build",
 		"lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
```
Thanks for working on this! Quick question - Is there a specific reason for setting the 'max-warnings' option to 0 in the 'lint' script?

taesungh replied:
This lint script came from the Vite template during project initialization, but to answer, setting the threshold ensures any warnings cause the `eslint` program to error as opposed to announcing the error but still finishing successfully (exit code 0). We want the lint check to fail on warnings accordingly. Refer to [`--max-warnings` in the ESLint CLI Reference](https://eslint.org/docs/latest/use/command-line-interface#--max-warnings) for more detail.

taesungh reviewed (approved):
Thanks for setting all of this up. Side note: pre-commit seems to not be maintained anymore, but we'll stick with this for now. We could maybe also look into using Turborepo in the future for simpler script delegation. Please squash during merge.

SM-dot commented on package.json:
Looks good! Just for clarification "pod operation backend"- what exactly are we planning to have for this?
